### PR TITLE
[ENH] PCA on sparse data

### PR DESCRIPTION
**Pull request by acopar**
##### Issue
Implements Issue #2255.
Use TruncatedSVD (LSA) on sparse data instead of PCA, since PCA does not work on sparse data. The difference is that TruncatedSVD truncates the vectors and does not center the data before calling svd. For dense data, functionality remains the same.

##### Description of changes
Workflow:
Corpus (Select bookexcerpts.tab) -> Bag of Words -> PCA
pca

- widgets/unsupervised/owpca.py: Data centering checkbox (true = PCA, false = TruncatedSVD)
- projection/pca.py: TruncatedSVD model
- tests: test_pca.py

##### Includes
- [X] Code changes
- [X] Tests
- [ ] Documentation

### CODE REVIEW

File: Orange/projection/pca.py [hunk 1]
```diff
@@ -152,6 +152,35 @@ def partial_fit(self, data):
...
+        self.max_components = max_components
+        self.params = vars()
+
+    def fit(self, X, Y=None):
+        params = self.params.copy()
+        if params["n_components"] is None and self.max_components is not None:
+            # shape of X after preprocessing
+            # Add -1 to avoid error in scikit fit_transform
+            # ValueError: n_components must be < n_features (strict)
+            params["n_components"] = min(min(X.shape) - 1, self.max_components)
```
Comment (kernc) [inline Orange/projection/pca.py]: Why, then, not have just `n_components` (and override it if shape too small)?

File: Orange/tests/test_pca.py [hunk 2]
```diff
@@ -84,6 +84,22 @@ def __ipca_test_helper(self, data, n_com, min_xpl_var):
         pc1_ipca = pca_model.partial_fit(data[1::2]).components_[0]
         self.assertAlmostEqual(abs(pc1_ipca.dot(pc1_pca)), 1, 4)
 
+    def test_truncated_svd(self):
+        data = self.ionosphere
+        self.__truncated_svd_test_helper(data, n_com=3, min_xpl_var=0.5)
+        self.__truncated_svd_test_helper(data, n_com=10, min_xpl_var=0.7)
+        self.__truncated_svd_test_helper(data, n_com=31, min_xpl_var=0.99)
+
+    def __truncated_svd_test_helper(self, data, n_com, min_xpl_var):
```
Comment (kernc) [inline Orange/tests/test_pca.py]: it's ok to use full, expressive variable names.

File: Orange/widgets/unsupervised/owpca.py [hunk 3]
```diff
@@ -42,6 +42,7 @@ class OWPCA(widget.OWWidget):
     auto_update = settings.Setting(True)
     auto_commit = settings.Setting(True)
     normalize = settings.Setting(True)
+    centralize = settings.Setting(True)
```
Comment (kernc) [inline Orange/widgets/unsupervised/owpca.py]: center

File: Orange/widgets/unsupervised/owpca.py [hunk 4]
```diff
@@ -125,6 +126,8 @@ def __init__(self):
         self.options_box = gui.vBox(self.controlArea, "Options")
         gui.checkBox(self.options_box, self, "normalize", "Normalize data",
                      callback=self._update_normalize)
+        gui.checkBox(self.options_box, self, "centralize", "Data centering",
```
Comment (kernc) [inline Orange/widgets/unsupervised/owpca.py]: "Center data"

File: Orange/widgets/unsupervised/owpca.py [hunk 5]
```diff
@@ -375,6 +385,26 @@ def _update_normalize(self):
...
+            self._pca_projector = PCA(max_components=MAX_COMPONENTS)
+            self._pca_projector.component = self.ncomponents
+            self._pca_preprocessors = PCA.preprocessors
+        else:
+            # Equivalent to PCA, but without data centering
+            self._pca_projector = TruncatedSVD(max_components=MAX_COMPONENTS)
+            self._pca_projector.component = self.ncomponents
+            self._pca_preprocessors = TruncatedSVD.preprocessors
+
+    def _update_centralize(self):
```
Comment (kernc) [inline Orange/widgets/unsupervised/owpca.py]: _update_center

File: Orange/widgets/unsupervised/owpca.py [hunk 6]
```diff
@@ -196,9 +200,15 @@ def set_data(self, data):
...
-                self.clear_outputs()
-                return
+                # PCA does not support sparse data
+                # Falling back to TruncatedSVD aka LSA
+                self.centralize = False
+                self._update_centralize()
+            else:
+                # Revert to PCA for non-sparse data
+                self.centralize = True
+                self._update_centralize()
```
Comment (kernc) [inline Orange/widgets/unsupervised/owpca.py]: put `_update_centralize()` call outside condition.

File: Orange/widgets/unsupervised/owpca.py [hunk 7]
```diff
@@ -375,6 +385,26 @@ def _update_normalize(self):
...
+            self._pca_projector.component = self.ncomponents
+            self._pca_preprocessors = PCA.preprocessors
+        else:
+            # Equivalent to PCA, but without data centering
+            self._pca_projector = TruncatedSVD(max_components=MAX_COMPONENTS)
+            self._pca_projector.component = self.ncomponents
+            self._pca_preprocessors = TruncatedSVD.preprocessors
+
+    def _update_centralize(self):
+        if self.centralize and self.data and self.data.is_sparse():
```
Comment (kernc) [inline Orange/widgets/unsupervised/owpca.py]: `self.data is not None`

**Comment by acopar** [comment]
Superseeded by pull request #2313 due to broken rebase structure.

**Comment by kernc** [comment]
The way you fix this in future is:
```bash
git fetch upstream master
git rebase --interactive upstream/master

# In the interactive editor that pops up, remove the commits
# that you obtained (and rewrote) from elsewhere

git push --force
```

**Pull request closed without merging**